### PR TITLE
improved indexer interaction with other subsystem

### PR DESCRIPTION
**Pull request by AdleyJ**
added isFeederAtSetpoint method
made intake work whenever there is no note in the robot
generally improved interaction between indexer and other subsystems

### CODE REVIEW

File: src/main/java/org/team1540/robot2024/commands/indexer/PrepareIndexerForShooter.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Override
+    public void initialize() {
+        indexer.setFeederVelocity(1200);
+        indexer.setIntakePercent(-0.5);
+    }
+
+    @Override
+    public boolean isFinished() {
+        return indexer.isFeederAtSetpoint() && !indexer.isNoteStaged();
```
Comment (rutmanz) [inline src/main/java/org/team1540/robot2024/commands/indexer/PrepareIndexerForShooter.java:23]: Intake will keep backing up until feeder is at speed, which is not desired behavior

File: src/main/java/org/team1540/robot2024/commands/indexer/StageTrampCommand.java [hunk 2]
```diff
@@ -14,12 +14,10 @@ public StageTrampCommand(Tramp tramp, Indexer indexer) {
         this.indexer = indexer;
         addRequirements(tramp, indexer);
     }
-
     @Override
     public void initialize() {
         tramp.setPercent(0.5);
         indexer.setFeederVelocity(-600);
-        indexer.setIntakePercent(0.5);
```
Comment (rutmanz) [inline src/main/java/org/team1540/robot2024/commands/indexer/StageTrampCommand.java:22]: Isn't this needed to actually push the note into the feeder to direct?

File: src/main/java/org/team1540/robot2024/subsystems/indexer/Indexer.java [hunk 3]
```diff
@@ -52,12 +52,21 @@ public Command feedToShooter() {
...
 
-    // TODO: Add method to check if feeder is spun up
+    public boolean isFeederAtSetpoint() {
+        return Math.abs(inputs.feederVelocityError) < VELOCITY_ERR_TOLERANCE_RPM;
+//        return MathUtil.isNear(inputs.setpointRPM, inputs.feederVelocityRPM, VELOCITY_ERR_TOLERANCE_RPM);
+    }
 
     public void stopFeeder() {
-        io.setFeederVoltage(0);
+        io.setFeederVoltage(0); // shouldn't stopFeeder probably set its velocity?
```
Comment (rutmanz) [inline src/main/java/org/team1540/robot2024/subsystems/indexer/Indexer.java:61]: I don't want it to actively move towards zero velocity, I want it to stop trying to move, brake mode will take care of the rest

**Review by rutmanz — CHANGES_REQUESTED** [review]

File: src/main/java/org/team1540/robot2024/commands/indexer/StageTrampCommand.java (hunk 2 above)
Comment (AdleyJ) [reply src/main/java/org/team1540/robot2024/commands/indexer/StageTrampCommand.java:22]: Simon had said that the feeder will be touching the note while it is in the indexer. Would that not be enough to push it into the tramp?

Comment (rutmanz) [reply src/main/java/org/team1540/robot2024/commands/indexer/StageTrampCommand.java:22]: If the note is being contacted by the intake belts (which it is here), having the intake not spin will actively oppose the motion that we want the note to take.

File: src/main/java/org/team1540/robot2024/commands/indexer/OuttakeCommand.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package org.team1540.robot2024.commands.indexer;
+
+import edu.wpi.first.wpilibj2.command.Command;
+import org.team1540.robot2024.subsystems.indexer.Indexer;
+
+public class OuttakeCommand extends Command {
```
Comment (rutmanz) [inline src/main/java/org/team1540/robot2024/commands/indexer/OuttakeCommand.java:6]: I would not call this outtake command if it's not actually ejecting. Maybe "StageIndexerShot" or something. This is also simple enough to be a `FunctionalCommand` if you want to do that

**Review by rutmanz — CHANGES_REQUESTED** [review]

File: src/main/java/org/team1540/robot2024/commands/shooter/ShootSequence.java [hunk 5]
```diff
@@ -1,17 +1,27 @@
...
 
 public class ShootSequence extends SequentialCommandGroup {
     public ShootSequence(Shooter shooter, Indexer indexer) {
         addCommands(
                 Commands.parallel(
-                        indexer.feedToShooter(),
-                        new PrepareShooterCommand(shooter)
+                        new PrepareShooterCommand(shooter),
+                        Commands.sequence(
+                                new FunctionalCommand(
```
Comment (rutmanz) [inline src/main/java/org/team1540/robot2024/commands/shooter/ShootSequence.java:16]: yeah I suppose this works, I was thinking more along the lines of a factory method (return this from something like `indexer.commandNoteBackout()`) but I suppose this is fine

**Review by rutmanz — APPROVED** [review]

**Pull request merged**